core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1013 @@
+};
+
+macro_rules! implement_per_thing {
+	($name:ident, $test_mod:ident, [$($test_units:tt),+], $max:tt, $type:ty, $upper_type:ty, $title:expr $(,)?) => {
+		/// A fixed point representation of a number between in the range [0, 1].
+		///
+		#[doc = $title]
+		#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug, Ord, PartialOrd))]
+		#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq)]
+		pub struct $name($type);
+
+		impl $name {
+			/// Nothing.
+			pub fn zero() -> Self { Self(0) }
+
+			/// `true` if this is nothing.
+			pub fn is_zero(&self) -> bool { self.0 == 0 }
+
+			/// Everything.
+			pub fn one() -> Self { Self($max) }
+
+			/// consume self and deconstruct into a raw numeric type.
+			pub fn deconstruct(self) -> $type { self.0 }
+
+			/// return the scale at which this per-thing is working.
+			pub const fn accuracy() -> $type { $max }
+
+			/// From an explicitly defined number of parts per maximum of the type.
+			///
+			/// This can be called at compile time.
+			pub const fn from_parts(parts: $type) -> Self {
+				Self([parts, $max][(parts > $max) as usize])
+			}
+
+			/// Converts from a percent. Equal to `x / 100`.
+			///
+			/// This can be created at compile time.
+			pub const fn from_percent(x: $type) -> Self {
+				Self([x, 100][(x > 100) as usize] * ($max / 100))
+			}
+
+			/// Return the product of multiplication of this value by itself.
+			pub fn square(self) -> Self {
+				// both can be safely casted and multiplied.
+				let p: $upper_type = self.0 as $upper_type * self.0 as $upper_type;
+				let q: $upper_type = $max as $upper_type * $max as $upper_type;
+				Self::from_rational_approximation(p, q)
+			}
+
+			/// Converts a fraction into `Permill`.
+			#[cfg(feature = "std")]
+			pub fn from_fraction(x: f64) -> Self { Self((x * ($max as f64)) as $type) }
+
+			/// Approximate the fraction `p/q` into a per-thing fraction. This will never overflow.
+			///
+			/// The computation of this approximation is performed in the generic type `N`. Given
+			/// `M` as the data type that can hold the maximum value of this per-thing (e.g. u32 for
+			/// perbill), this can only work if `N == M` or `N: From<M> + TryInto<M>`.
+			pub fn from_rational_approximation<N>(p: N, q: N) -> Self
+				where N: Clone + Ord + From<$type> + TryInto<$type> + ops::Div<N, Output=N>
+			{
+			// TODO: double check with @thiolliere the new bound. If your fraction type has a max
+			// that fits in u64, then you cannot build a rational from it that is `2u8 / 4u8` So
+			// using `SimpleArithmetic` here was okay as long as we don't add a new fraction type
+			// which is embedded in a u64. With perquintill we have to change this.
+				// q cannot be zero.
+				let q = q.max((1 as $type).into());
+				// p should not be bigger than q.
+				let p = p.min(q.clone());
+
+				let factor = (q.clone() / $max.into()).max((1 as $type).into());
+
+				// q cannot overflow: (q / (q/$max)) < $max. p < q hence p also cannot overflow.
```

Comment:
you are right that my comment is wrong but it can still not overflow. Look at this example: 

```rust
let max = 1_000_000_000_u32;
Perbill::from_rational_approximation(10u32, max + max - 1);
// p 10 / q 1999999999 / factor 1
// pr 10 / qr 1999999999
Perbill::from_rational_approximation(10u32, max + max);
// p 10 / q 2000000000 / factor 2
// pr 5 / qr 1000000000
Perbill::from_rational_approximation(10u32, max + max + 1);
// p 10 / q 2000000001 / factor 2
// pr 5 / qr 1000000000
```

q_reduce can exceed $max, but it can never exceed 2 * $max. Nonetheless a great catch and I should  update this assertion:

https://github.com/paritytech/substrate/blob/14fe8ac16d142e5c70aafcc1cd990b04e7374755/core/sr-primitives/src/sr_arithmetic.rs#L205

to `2 * $max`